codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -296,6 +296,7 @@ final DefaultPropertyKey verifyKey(PropertyKey key) {
         private int totalChildWeights;
         private int prioritizableForTree = 1;
         private boolean resetSent;
+        private boolean headerSent;
```

Comment:
This problem seems to stem from the remote flow controller queue. Does it make sense to move the responsibility there?  For example:

```java
public interface Http2RemoteFlowController extends Http2FlowController {
  public boolean isInitialHeadersQueued(Stream stream);
}

interface FlowControlled {
  public boolean isInitialHeaders();
}
```

When we create `FlowControlledHeaders` we need to pass in a `boolean` for is initial headers. `FlowControlledBase` can just return `false` by default. This way we shouldn't need additional interfaces/state on the Stream object.  WDYT?

Only thing I'm not sure about is if we need to treat push promise different https://github.com/netty/netty/pull/4502#discussion_r45893487.